File: solutions/2026-05-31/python/2026-05-31-destroying-asteroids.py
```python
# Problem Summary: Determine if a planet of given initial mass can destroy all asteroids by colliding with them in some order.
# Link: https://leetcode.com/problems/destroying-asteroids/
# Approach:
# The key insight here is that to maximize the planet's mass gain and thus its ability to destroy larger asteroids,
# we should always collide with the smallest available asteroid first. This is a greedy approach.
# By destroying smaller asteroids first, we increase the planet's mass as much as possible with the least resistance,
# making it more likely to be able to destroy larger asteroids later.
# Therefore, the optimal strategy is to sort the asteroids by mass in ascending order and then iterate through them,
# colliding with each one if the planet's current mass is sufficient. If at any point the planet's mass is less than
# the current asteroid's mass, it means the planet cannot destroy all asteroids, and we return False.
# If the planet successfully destroys all asteroids, we return True.
#
# Time Complexity Analysis:
# Sorting the asteroids takes O(N log N) time, where N is the number of asteroids.
# Iterating through the sorted asteroids takes O(N) time.
# Therefore, the overall time complexity is dominated by sorting, which is O(N log N).
#
# Space Complexity Analysis:
# Sorting the array might require O(log N) or O(N) space depending on the sorting algorithm implementation (e.g., Timsort in Python).
# The space used for variables is constant.
# Thus, the space complexity is O(N) or O(log N) due to sorting.

import sys

class Solution:
    def asteroidsDestroyed(self, mass: int, asteroids: list[int]) -> bool:
        """
        Determines if a planet can destroy all asteroids by colliding with them in an optimal order.

        Args:
            mass: The initial mass of the planet.
            asteroids: A list of asteroid masses.

        Returns:
            True if all asteroids can be destroyed, False otherwise.
        """

        # Sort the asteroids in ascending order of mass.
        # This is crucial for the greedy approach: destroying smaller asteroids first maximizes the planet's mass gain.
        asteroids.sort()

        # Initialize the current mass of the planet with its initial mass.
        current_mass = mass

        # Iterate through the sorted asteroids.
        for asteroid_mass in asteroids:
            # Check if the planet's current mass is sufficient to destroy the current asteroid.
            if current_mass >= asteroid_mass:
                # If sufficient, destroy the asteroid and increase the planet's mass.
                current_mass += asteroid_mass
            else:
                # If the planet's mass is less than the asteroid's mass, it cannot destroy this asteroid.
                # Therefore, it cannot destroy all asteroids. Return False.
                return False

        # If the loop completes without returning False, it means the planet has successfully destroyed all asteroids.
        return True

# Example Usage:
if __name__ == "__main__":
    solution = Solution()

    # Example 1
    mass1 = 10
    asteroids1 = [3, 9, 19, 5, 21]
    print(f"Input: mass = {mass1}, asteroids = {asteroids1}")
    print(f"Output: {solution.asteroidsDestroyed(mass1, asteroids1)}")  # Expected: True

    # Example 2
    mass2 = 5
    asteroids2 = [4, 9, 23, 4]
    print(f"Input: mass = {mass2}, asteroids = {asteroids2}")
    print(f"Output: {solution.asteroidsDestroyed(mass2, asteroids2)}")  # Expected: False

    # Additional Test Case: Large masses, needs careful sorting
    mass3 = 100
    asteroids3 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    print(f"Input: mass = {mass3}, asteroids = {asteroids3}")
    print(f"Output: {solution.asteroidsDestroyed(mass3, asteroids3)}") # Expected: True

    # Additional Test Case: A single large asteroid that cannot be destroyed
    mass4 = 50
    asteroids4 = [100]
    print(f"Input: mass = {mass4}, asteroids = {asteroids4}")
    print(f"Output: {solution.asteroidsDestroyed(mass4, asteroids4)}") # Expected: False

    # Additional Test Case: Many small asteroids that sum up to enough mass
    mass5 = 1
    asteroids5 = [1, 1, 1, 1, 1, 10]
    print(f"Input: mass = {mass5}, asteroids = {asteroids5}")
    print(f"Output: {solution.asteroidsDestroyed(mass5, asteroids5)}") # Expected: True

```

File: solutions/2026-05-31/cpp/2026-05-31-destroying-asteroids.cpp
```cpp
// Problem: Destroying Asteroids
// Link: https://leetcode.com/problems/destroying-asteroids/
// Summary: Determine if a planet can destroy all asteroids by colliding with them in some order.
//
// Approach:
// The key insight is that to maximize the planet's chances of destroying all asteroids,
// it should always collide with the smallest possible asteroid first. This greedy strategy
// ensures that the planet's mass increases as quickly as possible, making it more likely
// to be able to destroy larger asteroids later.
//
// 1. Sort the asteroids array in ascending order.
// 2. Iterate through the sorted asteroids. For each asteroid:
//    a. If the current planet mass is less than the asteroid's mass, the planet is destroyed,
//       and we return false.
//    b. If the planet's mass is greater than or equal to the asteroid's mass, the asteroid
//       is destroyed, and the planet's mass increases by the asteroid's mass.
// 3. If the loop completes without the planet being destroyed, it means all asteroids
//    could be destroyed, so return true.
//
// Time Complexity: O(N log N), where N is the number of asteroids. This is dominated by
// the sorting step. The iteration is O(N).
// Space Complexity: O(log N) or O(N) depending on the sorting algorithm used by the
// standard library's sort function. If an in-place sort is used, it could be O(log N)
// for recursion stack. If a copy is made, it could be O(N).

#include <vector>
#include <algorithm>

class Solution {
public:
    bool asteroidsDestroyed(int mass, std::vector<int>& asteroids) {
        // Sort the asteroids in ascending order to apply the greedy strategy.
        // This ensures we always try to destroy the smallest asteroids first,
        // maximizing the planet's mass gain.
        std::sort(asteroids.begin(), asteroids.end());

        // We need to use a long long for the planet's mass because the sum of masses
        // can exceed the range of an int.
        long long current_mass = mass;

        // Iterate through the sorted asteroids.
        for (int asteroid_mass : asteroids) {
            // If the current planet's mass is less than the asteroid's mass,
            // the planet cannot destroy this asteroid and will be destroyed itself.
            if (current_mass < asteroid_mass) {
                return false; // Planet destroyed
            }
            // If the planet's mass is sufficient, it destroys the asteroid and gains its mass.
            current_mass += asteroid_mass;
        }

        // If the loop completes, it means the planet successfully destroyed all asteroids.
        return true;
    }
};
```